Automattic/jetpack
# Connection: Select One of Multiple Blog Tokens

In certain situations, it is useful to:
* Not rely on the database for the blog token, and
* Allow multiple blog tokens per site (for secret rotation, for example).

### Changes proposed in this Pull Request:
* Allow the blog token to be set via a new `JETPACK_BLOG_TOKEN` constant.
* Support multiple blog tokens in `Jetpack_Data::get_access_token()` by returning either the first or the one matching the provided `$token_key`.
* Never call `Jetpack_Options::get_option( 'blog_token' )` directly. Switch to `Jetpack_Data::get_access_token()`
* Since the blog token is used to sign Jetpack Comments iframes, tweak the iframe's parameters to send and receive `$token_key` hints about which blog token to use.

### Testing instructions:

#### Automated Tests

`phpunit --filter=WP_Test_Jetpack_Data`

#### Manual Tests

1. Get the blog token value by doing:
   `wp jetpack options get blog_token`
2. In your wp-config.php file:
   `define( 'JETPACK_BLOG_TOKEN', '{that blog token value you just found}' );`
3. `wp jetpack options delete blog_token`
   You now have a site with no stored blog token. Instead, the Jetpack will use the blog token you defined in the `JETPACK_BLOG_TOKEN` constant.
4. Go to wp-admin/ → Tools → Site Health → Info
5. Click "Jetpack"
6. Scroll down and see that "Blog Public Key" and "User Public Key" are set and that the Blog Public Key matches the token you found in step 1.


##### Testing XML-RPC & JSON API
1. Blog Token Test:
   `curl -i 'https://public-api.wordpress.com/rest/v1/sites/{YOUR_JETPACK_SITE}'`
2. User Token Test:
   1. Go to https://developer.wordpress.com/docs/api/console/
   2. Make sure you're logged in to WordPress.com as a user that corresponds to a connected user on the Jetpack site.
   3. Select "WP.COM API", "v1.1", "GET"
   4. `/sites/{YOUR_JETPACK_SITE}`
   5. Hit enter
3. For both, you should see some a JSON response of site data.

##### Testing OAuth2 Connections
1. Create an OAuth2 app and make sure you can obtain an OAuth2 token for a Jetpack site.

This is super annoying to test unless you already have a working app. Automatticians can find one at p9dueE-MU-p2.

##### Testing Jetpack Comments
1. Make sure you have Jetpack Comments turned on (wp-admin/ → Jetpack → Settings → Discussion → "Let readers use WordPress.com…"
2. Go to a post's permalink.
3. See that the Comments iframe displays correctly.
4. Submit a comment.
5. See that the comment ends up in the Jetpack site's comments at wp-admin/ → Comments.

### Proposed changelog entry for your changes:
None required.
